wallet/src/main/scala/org/bitcoins/wallet/Wallet.scala

```diff
@@ -44,23 +61,82 @@ sealed abstract class Wallet extends UnlockedWalletApi with BitcoinSLogger {
   /**
     * @inheritdoc
     */
-  override def addUtxo(utxo: UTXOSpendingInfo): Future[WalletApi] = ???
+  def addUtxo(
+      transaction: Transaction,
+      vout: UInt32): Future[Either[AddUtxoError, WalletApi]] = {
+    val voutIndexOutOfBounds = transaction.outputs.length >= vout.toInt
+
+    val outputE: Either[AddUtxoError, TransactionOutput] =
+      if (voutIndexOutOfBounds) Left(VoutIndexOutOfBounds)
+      else
+        Right(transaction.outputs(vout.toInt))
+
+    val outPointE: Either[AddUtxoError, TransactionOutPoint] =
+      if (voutIndexOutOfBounds) Left(VoutIndexOutOfBounds)
+      else Right(TransactionOutPoint(transaction.txId, vout))
+
+    val addressEitherF: Future[Either[AddUtxoError, AddressDb]] = {
+      val nestedAddressE = for {
+        output <- outputE
+        address <- BitcoinAddress
+          .fromScriptPubKey(output.scriptPubKey, networkParameters)
+          .toEither
+          .left
+          .map(_ => BadSPK)
+      } yield
+        addressDAO.findAddress(address).map {
+          case Some(addr) => Right(addr)
+          case None       => Left(AddressNotFound)
+        }
+      EitherUtil.flattenFutureE(nestedAddressE)
+    }
+
+    addressEitherF.flatMap { addressE =>
+      val nestedWalletE: Either[
+        AddUtxoError,
+        Future[Either[AddUtxoError, WalletApi]]] =
+        for {
+          output <- outputE
+          outPoint <- outPointE
+          address <- addressE
+        } yield {
+          val utxo =
+            UTXOSpendingInfoDb(None, outPoint, output, address.path, None, None)
+          utxoDAO.create(utxo).map(_ => Right(this.asInstanceOf[WalletApi]))
+        }
+      EitherUtil.flattenFutureE(nestedWalletE)
+    }
+  }
```

Comment:
I think this code becomes much cleaner if you change `Either[AddUtxoError, T] -> Try[T]` and `Future[Either[AddUtxoError, T]] -> Future[T]` since `Future`s can fail (They're little more than (mutable) wrappers for `Try[Option[_]]`)

Reply:
Two issues with that:
1) `Try[T]` lacks well defined errors. With `AddUtxoError` being an ADT, consumers can be sure they handle (or at least are aware of) every error
2) `Future` doesn't force users to handle the error case, and the error is also untyped (like `Try`)

Although I agree that handling nested `Either`s and `Future`s isn't ideal. Something like `EitherT` from Cats would handle this better: https://typelevel.org/cats/datatypes/eithert.html

So eventually we should implement some of that functionality, for cases like this

Reply:
In this case, what do you think of creating a wrapper type specific to our use case? I.e. instead of using `Try`, creating an ADT like it with a success case and a failure case where the failure wraps an `AddUtxoError`? I always get super confused when looking at maps on `Either` because it isn't obvious to me which one is the "success" case.

Reply:
I ended up doing `Either` internally, and ADT for public facing API. What do you think about that?

Reply:
Also did this in `initializeWithEntropy`